src/components/select/select-list/select-list.component.js

```diff
@@ -93,62 +115,89 @@ const SelectList = React.forwardRef(
+    // as that isn't absolutely guaranteed to never rerun when dependencies haven't changed.
+    const setChildIds = () => {
+      childIdsRef.current = React.Children.map(children, () => guid());
+    };
+
+    if (childIdsRef.current?.length !== React.Children.count(children)) {
+      setChildIds();
+    }
+
+    const childIds = childIdsRef.current;
+
+    const childrenList = useMemo(() => React.Children.toArray(children), [
       children,
     ]);
 
-    const childrenWithListProps = useMemo(
+    const optionChildrenList = useMemo(
       () =>
-        React.Children.map(children, (child, index) => {
-          if (!child || (child.type !== Option && child.type !== OptionRow)) {
-            return child;
-          }
+        childrenList.filter(
+          (child) => child.type === Option || child.type === OptionRow
+        ),
+      [childrenList]
+    );
 
-          const newProps = {
-            index,
-            id: childIds[index],
-            onSelect: handleSelect,
-            hidden: isLoading && React.Children.count(children) === 1,
-            ref: optionRefList[index],
-          };
+    const { measureElement } = virtualizer;
 
-          return React.cloneElement(child, newProps);
-        }),
-      [children, handleSelect, isLoading, optionRefList, childIds]
-    );
+    const measureCallback = (element) => {
+      // need a guard to prevent crash with too many rerenders when closing the list
+      if (isOpen) {
+        measureElement(element);
+      }
+    };
 
-    const childrenList = useMemo(
-      () => React.Children.toArray(childrenWithListProps),
-      [childrenWithListProps]
-    );
+    const renderedChildren = items.map((item) => {
+      const { index, start } = item;
+      const child = childrenList[index];
 
-    const lastOptionIndex = useMemo(() => {
-      let lastIndex = 0;
+      if (!child) {
+        return child;
+      }
 
-      childrenList.forEach((element, index) => {
-        if (element.type === Option || element.type === OptionRow) {
-          lastIndex = index;
-        }
-      });
+      const optionChildIndex = optionChildrenList.indexOf(child);
+      const isOption = optionChildIndex > -1;
+
+      const newProps = {
+        index,
+        id: childIds[index],
+        onSelect: handleSelect,
+        hidden: isLoading && childrenList.length === 1,
+        // these need to be inline styles rather than implemented in styled-components to avoid it generating thousands of classes
+        style: {
+          transform: `translateY(${start}px)`,
+        },
+        "aria-setsize": isOption ? optionChildrenList.length : undefined,
+        "aria-posinset": isOption ? optionChildIndex + 1 : undefined,
+        // needed to dynamically compute the size
+        ref: measureCallback,
+        "data-index": index,
+      };
 
+      return React.cloneElement(child, newProps);
+    });
+
+    const lastOptionIndex = useMemo(() => {
```

Comment:
Might be worth using the function provided by `lodash`. It's simple enough that I'm happy to leave this as is if you'd prefer